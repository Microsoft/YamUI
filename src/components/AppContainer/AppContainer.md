### Notes for use

You should wrap your entire app in an `AppContainer` component. It renders a `Fabric` component and some YamUI baseline styles, together this manages our user-scalable font sizing scheme, language-specific font stack, and some special styling for accessible keyboard tabbing.

### Examples

Basic
```js { "props": { "data-description": "basic" } }
<AppContainer>
  Wrap the root of your application in AppContainer.
</AppContainer>
```
