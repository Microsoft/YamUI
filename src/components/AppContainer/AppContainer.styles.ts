/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';
import { textColors, fontWeights } from '../../util/styles/fonts';

export interface AppContainerClassNameProps {
  theme: ITheme;
}

export const getClassNames = memoizeFunction((props: AppContainerClassNameProps) => {
  const { theme } = props;
  const font = theme.fonts.medium;

  return mergeStyleSets({
    root: {
      color: textColors(theme).primary,
      fontSize: font.fontSize,
      lineHeight: font.lineHeight,
      fontWeight: fontWeights.normal,

      selectors: {
        ':global(html)': {
          /* Sets 1.0rem to 10px */
          fontSize: '62.5%',
          boxSizing: 'border-box',
        },
        ':global(html, body)': {
          margin: 0,
          padding: 0,
        },
        ':global(*, *::before, *::after)': {
          boxSizing: 'inherit',
        },
      },
    },
  });
});
