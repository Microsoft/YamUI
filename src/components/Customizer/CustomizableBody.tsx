/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { CustomizableComponentProps, defaultTheme, customizable } from './Customizer.types';
import { join } from '../../util/classNames';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { getClassNames } from './Customizer.styles';

export class Body extends React.Component<CustomizableComponentProps & NestableBaseComponentProps> {
  public render() {
    const { children, className, theme = defaultTheme } = this.props;
    const classNames = getClassNames({ theme });
    const rootClassNames = join(['y-customizable-body', classNames.root, className]);

    return <div className={rootClassNames}>{children}</div>;
  }
}

@customizable('Body', ['theme'])
export default class CustomizableBody extends Body {}
