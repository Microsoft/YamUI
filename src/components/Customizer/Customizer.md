### Examples

Button with theme:

```js { "props": { "data-description": "basic" } }
const { createTheme, getTheme } = require('./');
const Button = require('../Button').default;

const grayTheme = createTheme({
  palette: {
    themeDark: '#333',
    neutralDark: '#666',
    themeLighter: '#999',
  },
  fonts: {
    smallPlus: {
      fontSize: '1.8rem',
      lineHeight: '2.4rem',
    },
  }
});

<div>
  <Button text="Normal" />{' '}
  <Customizer settings={{ theme: grayTheme }}>
    <Button text="Big Gray Theme" />
  </Customizer>
</div>;
```
