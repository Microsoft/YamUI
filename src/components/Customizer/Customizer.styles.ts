/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';

export interface CustomizerClassNameProps {
  theme: ITheme;
}

export const getClassNames = memoizeFunction((styleProps: CustomizerClassNameProps) => {
  const { theme } = styleProps;
  const { fontSize, lineHeight, fontFamily, fontWeight } = theme.fonts.medium;

  return mergeStyleSets({
    root: {
      backgroundColor: theme.semanticColors.bodyBackground,
      color: theme.semanticColors.bodyText,
      fontSize,
      lineHeight,
      fontFamily,
      fontWeight,
    },
  });
});
