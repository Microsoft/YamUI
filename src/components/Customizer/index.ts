/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default, createTheme } from './Customizer';
export * from './Customizer';
