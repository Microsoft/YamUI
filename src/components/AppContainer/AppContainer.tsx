/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { Fabric } from 'office-ui-fabric-react/lib/Fabric';
import Customizer, { CustomizableComponentProps, defaultTheme } from '../Customizer';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { getClassNames, scopedSettings } from './AppContainer.styles';

/**
 * AppContainer sets some baseline visual and accessibility styling.
 */
export default class AppContainer extends React.Component<NestableBaseComponentProps & CustomizableComponentProps> {
  public render() {
    const { children, className, theme = defaultTheme } = this.props;
    const classNames = getClassNames();
    return (
      <Fabric theme={theme}>
        <Customizer settings={{ theme }} scopedSettings={scopedSettings}>
          <div className={join(['y-appContainer', classNames.root, className])}>{children}</div>
        </Customizer>
      </Fabric>
    );
  }
}
