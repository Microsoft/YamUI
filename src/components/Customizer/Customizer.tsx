/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { Customizer as FabricCustomizer, ICustomizerProps } from 'office-ui-fabric-react/lib/Utilities';
import CustomizableBody from './CustomizableBody';

/**
 * Used to inject a custom theme object into customizable components, overriding the global theme.
 */
export default class Customizer extends React.Component<ICustomizerProps & NestableBaseComponentProps> {
  public render() {
    const { children, className, ...customizerProps } = this.props;

    return (
      <FabricCustomizer {...customizerProps}>
        <CustomizableBody {...{ children, className }} />
      </FabricCustomizer>
    );
  }
}
