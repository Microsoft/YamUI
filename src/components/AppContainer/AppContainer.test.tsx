/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import AppContainer from '.';
import { darkTheme } from '../Customizer';

describe('<AppContainer />', () => {
  let component: ShallowWrapper<NestableBaseComponentProps>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<AppContainer>app container text</AppContainer>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with className', () => {
    beforeEach(() => {
      component = shallow(<AppContainer className="TEST_CLASSNAME">app container text</AppContainer>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with theme', () => {
    beforeEach(() => {
      component = shallow(<AppContainer theme={darkTheme}>app container text</AppContainer>);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
