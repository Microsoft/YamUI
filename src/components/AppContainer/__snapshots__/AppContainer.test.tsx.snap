// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppContainer /> with className matches its snapshot 1`] = `
<CustomizedFabric>
  <div
    className=
        y-appContainer
        TEST_CLASSNAME
        {
          color: #343a41;
          font-size: 1.5rem;
          font-weight: 400;
          line-height: 2.0rem;
        }
        html {
          font-size: 62.5%;
        }
  >
    paragraph text
  </div>
</CustomizedFabric>
`;

exports[`<AppContainer /> with default options matches its snapshot 1`] = `
<CustomizedFabric>
  <div
    className=
        y-appContainer
        {
          color: #343a41;
          font-size: 1.5rem;
          font-weight: 400;
          line-height: 2.0rem;
        }
        html {
          font-size: 62.5%;
        }
  >
    paragraph text
  </div>
</CustomizedFabric>
`;
